### ISSUE TITLE
Remote-join is broken when downstream is entered multiple times

### ISSUE THREAD

**Issue by jithine** [screwdriver-cd/screwdriver#2980]
**What happened**:

Remote join job is not triggered if the downstream pipeline is part of more than 1 remote trigger scenario

Pipeline demonstrating issue: https://cd.screwdriver.cd/pipelines/12907/events/784127 In this pipeline the last remote-join job `third` is not triggered. Only `second` job works 
<img width="797" alt="Screenshot 2024-01-11 at 3 19 20 PM" src="https://github.com/screwdriver-cd/screwdriver/assets/193126/fda56be7-2bcd-497a-84f0-25b3592cede9">

In this instance the downstream pipelines are entered twice as part of remote-join workflow. 

**What you expected to happen**:

1. Remote-Join job `third` should have been triggered.
2. downstream job should have displayed proper trigger during second remote-trigger. In current scenario, the 2nd remote-trigger happens in the event where first remote-trigger happened, but it's upstream is not displayed in workflow graph
3. Remote join should have worked whether `~` is used or not. In this instance I wasn't able to get any remote-trigger to work without `~` in downstream job Eg event https://cd.screwdriver.cd/pipelines/12908/events/784107
**How to reproduce it**:

https://github.com/jithine/remote-join-upstream/blob/main/screwdriver.yaml
https://github.com/jithine/remote-join-downstream-l2-1/blob/main/screwdriver.yaml
https://github.com/jithine/remote-join-downstream-l2-2/blob/main/screwdriver.yaml
https://github.com/jithine/remote-join-downstream-l2-3/blob/main/screwdriver.yaml

**Comment by jithine** [comment]
FYI @kumada626 @y-oksaku

**Comment by y-oksaku** [comment]
## Problem 1
Several builds in the downstream were not triggered by the upstream build. (eg. second and third in [this event](https://cd.screwdriver.cd/pipelines/12910/events/784126))
Consequently, the join list of the third job in the upstream was not satisfied.

This bug is caused by a 'return' within the for-loop here.
https://github.com/screwdriver-cd/screwdriver/blob/3170afbb20eee4d6b6936e9332a279811054e165/plugins/builds/index.js#L1087-L1096

## Problem 3
This problem is similar to https://github.com/screwdriver-cd/screwdriver/pull/2964 .

Only `~sd@` is triggered in [eventFactory.createBuilds](https://github.com/screwdriver-cd/models/blob/master/lib/eventFactory.js#L341).
So `sd@` is not triggered when create downstream event at here.
https://github.com/screwdriver-cd/screwdriver/blob/3170afbb20eee4d6b6936e9332a279811054e165/plugins/builds/index.js#L1119-L1135

I think it would work if we add trigger processing in the same way.

**Comment by jithine** [comment]
Thanks for checking @y-oksaku Problem 1 is our immediate priority please.

**Comment by jithine** [comment]
@y-oksaku we've ended up rolling back the change to unblock internal users. Once you have a fix, pls go ahead and undo these reverts.

**Comment by y-oksaku** [comment]
@jithine 
OK, but please note that after revert, `Y2` and `bar` in the following are not triggered.

upstream(123)

```yaml
jobs:
  hub:
    requires: [ ~commit, ~pr ]
  test:
    requires: []
  test2:
    requires: []
  Y2:
    requires: [ ~sd@12926:foo, test, test2 ]
```

downstream(456)

```yaml
jobs:
  foo:
    requires: [ ~sd@123:hub ]
  bar:
    requires: [ sd@123:hub ]
```

**Comment by jithine** [comment]
@y-oksaku that's unfortunate, looks like we would end up breaking some customer's workflow. Once you have the fix ready for the workflow in this issue, let's immediately roll forward and bring back all the reverted changes.

**Comment by y-oksaku** [comment]
## Problem 2

<img src="https://github.com/screwdriver-cd/screwdriver/assets/43719835/88117cff-92f3-4efe-813c-6a7149d26db5" height="400px">

I agree that it would be better to display the 2nd upstream remote trigger on the downstream.
However, this display is shown in the [UI](https://github.com/screwdriver-cd/ui/blob/2831be488d5065ebd3a1c309af5d747fbdd7b015/app/utils/graph-tools.js#L282-L285) based on the [event.startFrom](https://github.com/screwdriver-cd/data-schema/blob/master/models/event.js#L47-L50) data, so we would need to either modify this column or add a new one.
This column is also used in the trigger process, so it would be beneficial to address this along with other trigger issues.

**Comment by jithine** [comment]
> @jithine OK, but please note that after revert, `Y2` and `bar` in the following are not triggered.
> 
> upstream(123)
> 
> ```yaml
> jobs:
>   hub:
>     requires: [ ~commit, ~pr ]
>   test:
>     requires: []
>   test2:
>     requires: []
>   Y2:
>     requires: [ ~sd@12926:foo, test, test2 ]
> ```
> 
> downstream(456)
> 
> ```yaml
> jobs:
>   foo:
>     requires: [ ~sd@123:hub ]
>   bar:
>     requires: [ sd@123:hub ]
> ```

@y-oksaku can you describe what is not working in this workflow ?
My analysis here
1. `bar` is not working because `~` is missing
2. Y2 should have triggered if either ` ~sd@12926:foo` runs or if `test` & `test2` runs.

**Comment by jithine** [comment]
> ## Problem 2
> <img alt="" height="400px" src="https://private-user-images.githubusercontent.com/43719835/298812471-88117cff-92f3-4efe-813c-6a7149d26db5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYyMzU1NDgsIm5iZiI6MTcwNjIzNTI0OCwicGF0aCI6Ii80MzcxOTgzNS8yOTg4MTI0NzEtODgxMTdjZmYtOTJmMy00ZWZlLTgxM2MtNmE3MTQ5ZDI2ZGI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI2VDAyMTQwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThjYmE4NjNjMTRlZjc0MmNmN2Y0MmVhOTM3MDU5NDA5M2QxNDAwMGMzNTMyMjdiYjU4YWQ3OGI0MjFmMjc1MzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DgxWqn6Ah2q5I9gwEIoe2AijZLFIUOb6FlFJ8zGLYI4">
> 
> I agree that it would be better to display the 2nd upstream remote trigger on the downstream. However, this display is shown in the [UI](https://github.com/screwdriver-cd/ui/blob/2831be488d5065ebd3a1c309af5d747fbdd7b015/app/utils/graph-tools.js#L282-L285) based on the [event.startFrom](https://github.com/screwdriver-cd/data-schema/blob/master/models/event.js#L47-L50) data, so we would need to either modify this column or add a new one. This column is also used in the trigger process, so it would be beneficial to address this along with other trigger issues.

@y-oksaku in this case, I think rather than attempting to display `green` on 2nd remote trigger, we should try to continue the graph by adding a graph `edge` from node `component` to node `~sd@12907@compent`

**Comment by y-oksaku** [comment]
@jithine 

> @y-oksaku can you describe what is not working in this workflow ? My analysis here
> 
> 1. `bar` is not working because `~` is missing, we should address this scenario for sure since all trigger conditions have been satisfied.

Yes. `~` is missing, so `bar` job is not triggered when the build `hub` completed.
This bug is not caused by PRs you reverted, so it is still occurring.

> 2. Y2 should have triggered if either ` ~sd@12926:foo` runs or if `test` & `test2` runs.

`Y2` is not triggered when downstream job `foo` completed.
(Sorry, there is a mistake: `[ ~sd@12926:foo, test, test2 ]` → `[ ~sd@456:foo, test1, test2 ]`)
This problem fixed in https://github.com/screwdriver-cd/screwdriver/pull/2964 but now it was reverted since this fix caused the main Problem 1.

**Comment by jithine** [comment]
@y-oksaku my test pipeline worked with latest API. So I think the main issue has been fixed. Should I separate out the problem 2 into a separate issue and close this ?
